### ISSUE TITLE
Add more descriptive parameter type specifications for custom commands

### ISSUE THREAD

**Issue by whooie** [nushell/nushell#6654]
When writing custom commands, I often find myself wanting the ability to specify that a command should take, specifically, a `list<float>` or a record with a certain set of fields, which could conceivably by written as something like `record<f1: float, f2: float>`. I know one could just use `describe` for this kind of thing, but that feels hacky and unnecessary in principle.

Relatedly, support for specifying what type the contents of `$in` should be (like how the various `math` functions are able to detect non-numerical piped inputs) would also be greatly appreciated. This feels less necessary, since one could always just use a helper function operating on `$in`, but I think both of these things would go a long way.

**Comment by fdncred** [comment]
Transferred to nushell since the issue wasn't about documentation.

**Comment by lordkekz** [comment]
I think this issue can could be closed since it's mostly implemented as of Nushell 0.83.1.
I have tested the current behavior, which I believe to be as requested with regards to the parameter type specifications:
```nu
def test_typed_list [param: list<int>] { echo $param | describe }

# Works; output: list<int>
test_typed_list [1 2 3]
# Error: nu::parser::parse_mismatch (expected int)
test_typed_list [a b c]

def test_typed_record [param: record<a: int, b: string>] { echo $param | describe }

# Works; output: record<a: int, b: string>
test_typed_record {a: 1, b: 'hello'}
# Error: nu::parser::type_mismatch (expected record<a: int, b: string>, found record<a: string, b: int>)
test_typed_record {a: 'oops', b: 42}

def test_typed_rest [...param: int] { echo $param | describe }

# Works; output: list<any>
test_typed_rest
# Works; output: list<int>
test_typed_rest 1 2 3
# Error: nu::parser::parse_mismatch (expected int)
test_typed_rest a b c
```
(Note that if `test_typed_rest` was changed to expect a string and we gave an int literal, it would automatically cast; see https://github.com/nushell/nushell/issues/9598#issuecomment-1627714179)

It also mostly works with input/output type hints:
```nu
def test_typed_list_pipe []: list<int> -> string { describe }

# Works; output: list<int>
[1 2 3] | test_typed_list_pipe
# Error: nu::parser::input_type_mismatch (command doesn't support string input)
[a b c] | test_typed_list_pipe

# There seems to be a parser bug with record signatures: (note that there was no such problem with lists)
# Error: nu::parser::unclosed_delimiter
#  × Unclosed delimiter.
#   ╭─[entry #223:1:1]
# 1 │ def test_typed_record_pipe []: record<a: int, b: string> -> string { describe }
#   ·                                ────┬────
#   ·                                    ╰── unclosed >
#   ╰────
def test_typed_record_pipe []: record<a: int, b: string> -> string { describe }

# Works, but lacks the ability to type check the record signatures
def test_typed_record_pipe []: record -> string { describe }
```

However, type specifications do not seem to be enforced at all if the `$in` variable is used:
```nu
# This *should* give an error, since test_typed_list expects a list<int> but the pipe resolves to a list<string>
# Current behavior: runs; output: list<string>
[a b c] | test_typed_list $in

def test_madly_typed [param: list<string>]: list<int> -> string { describe | print; echo $param | describe }

# Works; output: nothing and list<string>
test_madly_typed [a b c]
# Works; output: list<int> and list<string>
[1 2 3] | test_madly_typed [a b c]
# Error: nu::parser::input_type_mismatch (command doesn't support string input)
"test" | test_madly_typed [a b c]

# This *should* give an error because param expects list<string> but gets string, and also because
# command input expects list<int> but gets string.
# I also was surprised that we still get an input even though it's been passed as a parameter too.
# Runs; output: string and string
"test" | test_madly_typed $in
```

Apologies for the verbosity; I didn't even realize input/output type hints were a thing, and didn't find much documentation on them, so I figured this might be useful to others.

**Comment by fdncred** [comment]
Thanks!
